[stdlib] RandomAccessCollection conformance for standard integer Words

lorentey opened the pull request:
<!-- What's in this pull request? -->
This PR updates the definition of `Words` on all standard integer types (from `Int8` to `UInt64`) to implement `RandomAccessCollection`.

For integer types that fit in a single word, the custom `Words` collection is replaced with `CollectionOfOne<UInt>`. The original collection is still used by `Int64` and `UInt64` on 32-bit platforms, though, but it is updated to implement `RandomAccessCollection`.

Making `Int.Words` (or `UInt.Words`) conform to `RandomAccessCollection` reveals a type inference issue with `RandomAccessSlice.Indices`: it seems that in this case, `Int.Words.SubSequence.Indices` does not default to `DefaultRandomAccessIndices`, but rather `DefaultIndices`, which breaks `RandomAccessCollection` requirements. I worked around this by adding an explicit definition for `Indices` to all slice types, forcing the compiler to always use the correct type.

This PR ~~has negligible impact on~~ **slightly improves** stdlib compilation time.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

lorentey committed:
[stdlib] Use RandomAccessCollections as [U]Int*.Words

For integer types that fit in a word, Words is now a type alias
to CollectionOfOne. Others use a custom random access collection.

lorentey committed:
[stdlib] Set Indices explicitly on slices.

RandomAccessSlice.Indices otherwise gets defaulted to
DefaultIndices rather than DefaultRandomAccessIndices
when Int.Words is a random access collection.

lorentey commented:
cc @moiseev

moiseev commented:
Such a small change, such a huge impact! Thanks @lorentey!

moiseev commented:
@swift-ci Please test

moiseev commented:
@swift-ci Please smoke benchmark

lorentey commented:
I reported the type inference issue as https://bugs.swift.org/browse/SR-5609

moiseev commented:
@swift-ci Please test
The CI error seem to have been fixed now.

moiseev commented:
@swift-ci Please test

moiseev commented:
@swift-ci Please smoke benchmark

milseman commented:
@moiseev it's been about 2 weeks for this PR. What was the verdict?

moiseev commented:
@milseman The only reason I did not merge it is I did not have time to check its effects on binary size and compilation time. @lorentey maybe you can do it (now that we're on the same page on what to measure).

milseman commented:
Also, @lorentey or @moiseev, please update the Standard Library Programmer's Manual with how you do it and what to look for! I've only done a few ad-hoc runs of `cmpcodesize -a -s`, but I'm interested in how you do it.

lorentey commented:
Good news! The removal of all these custom collections speeds up compilation by a few seconds and saves about 78 kiB:

```
$ time ninja -C ./build/Ninja-ReleaseAssert/swift-macosx-x86_64 swift-stdlib-macosx
before: 121.06 real       238.98 user         8.62 sys
after:  117.19 real       232.01 user         8.01 sys
```

```
$ ./swift/utils/cmpcodesize/cmpcodesize.py -a ~/Git/Swift/build.master/Ninja-ReleaseAssert/swift-macosx-x86_64/lib/swift/macosx/x86_64/libswiftCore.dylib -- ~/Git/Swift/build/Ninja-ReleaseAssert/swift-macosx-x86_64/lib/swift/macosx/x86_64/libswiftCore.dylib 
                                   Section             Old             New  Percent
libswiftCore.dylib                  __text:        4530609         4451537     1.7%
    section                  __textcoal_nt:              0               0       - 
    section                        __stubs:           1230            1230     0.0%
    section                        __const:         188848          179696     4.8%
    section                      __cstring:          68908           68844     0.1%
    section                __objc_methname:           1543            1543     0.0%
    section                        __const:         188848          179696     4.8%
    section                   __objc_const:          16464           16464     0.0%
    section                         __data:         150304          149728     0.4%
    section                 __swift1_proto:              0               0       - 
    section                       __common:            192             192     0.0%
    section                          __bss:          74864           74784     0.1%
```

These numbers are for release builds (generated by `build-script -R`). Compilation timings follow the @moiseev Method as described in #11178. :-)

I'll look into updating the manual! However, I'll first have to track down a few issues: pip has trouble installing `cmpcodesize`, and there seems to be something nondeterministic about the compilation time benchmark. Measured values are sometimes off by as much as ~25 seconds, due to ninja apparently sometimes finding a different set of files to recompile. The relative results remain unaffected, but the difference is unsettling.

lorentey commented:
One questionable aspect of this PR that it removes the use of the custom `Words` collection for all but 32-bit builds. Therefore, the code implementing it would not receive much testing -- it only gets used on watchOS, I think.

If this is deemed a problem, a useful compromise would be to keep the custom collection for `UInt.Words`, and have the `Words` of all smaller integer types be a typealias to it, rather than `CollectionOfOne`. This would benchmark the largely same, but there would be more symmetry between 32- and 64-bit platforms. (Unfortunately `CollectionOfOne` would not get used.)

lorentey commented:
I liked the alternative idea above so much that I added a commit implementing a `DefaultWords<Integer>` type and setting it as a default for `FixedWidthInteger.Words`. (I hope this is okay to do without a round trip to swift-evolution. If not, we can always take it out.) 

This way, all standard integers have consistent `Words` types on all platforms.

Updated compilation time (vs master):
```
$ time ninja -C ./build/Ninja-ReleaseAssert/swift-macosx-x86_64 swift-stdlib-macosx
before: 121.06 real       238.98 user         8.62 sys
after:  118.03 real       233.47 user         8.24 sys
```

stdlib size:

```
$ ./swift/utils/cmpcodesize/cmpcodesize.py -a ~/Git/Swift/build.master/Ninja-ReleaseAssert/swift-macosx-x86_64/lib/swift/macosx/x86_64/libswiftCore.dylib -- ~/Git/Swift/build/Ninja-ReleaseAssert/swift-macosx-x86_64/lib/swift/macosx/x86_64/libswiftCore.dylib 
                                   Section             Old             New  Percent
libswiftCore.dylib                  __text:        4530609         4457681     1.6%
    section                  __textcoal_nt:              0               0       - 
    section                        __stubs:           1230            1230     0.0%
    section                        __const:         188848          180608     4.4%
    section                      __cstring:          68908           68908     0.0%
    section                __objc_methname:           1543            1543     0.0%
    section                        __const:         188848          180608     4.4%
    section                   __objc_const:          16464           16464     0.0%
    section                         __data:         150304          151328    -0.7%
    section                 __swift1_proto:              0               0       - 
    section                       __common:            192             192     0.0%
    section                          __bss:          74864           74784     0.1%
```

With the new commit, the PR is still a win, although by a slightly narrower margin.

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2204,6 +2206,56 @@ ${assignmentOperatorComment(x.operator, False)}
+///
+/// The words of negative values are in two's complement representation,
+/// regardless of the integer type's underlying implementation.
+public struct DefaultWords<Integer: FixedWidthInteger> : RandomAccessCollection {
```
`IntegerWordsCollection` perhaps? `Default` really only makes sense if we have more than one, which we don't. Or simply `IntegerWords<Storage>`?

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2204,6 +2206,56 @@ ${assignmentOperatorComment(x.operator, False)}
+      return (_value &>> Integer(_truncatingBits: shift))._lowWord
+    }
+  }
+}
```
I can't prove it, but implementing a custom iterator should be beneficial, as otherwise the default `IndexingIterator` will be used and it doesn't know that in most cases we have only one element, and thus will be suboptimal. We can benchmark iterating over integer words and then add this improvement in a separate PR to see whether it makes any difference.

lorentey replied:
Good idea; I'll do a quick microbenchmark on iterating over `words` with and without a custom iterator for the single-word case (after I work around the `_lowWord` issue).

moiseev commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -2204,6 +2206,56 @@ ${assignmentOperatorComment(x.operator, False)}
+  }
+}
+
+extension FixedWidthInteger where Words == DefaultWords<Self> {
```
I think that having the `typealias Words` and this conditional expression is unnecessarily complicated to track. Why not have them grouped together in the same non-conditional extension?

lorentey replied:
Ah, what a neat approach. 👍

lorentey commented:
Darn! `DefaultWords.subscript` relies on `_lowWord` having a custom implementation, so it doesn't make a good default for `FixedWidthInteger.Words`, and it should not be used for integer types outside of the standard library. This stuff is trickier than I thought!

Unfortunately, I see no easy way out of this, so it looks like I have to withdraw the last commit. Sorry for wasting our time. :(

The one remaining option to fix the inconsistencies between 32/64-bit platforms is to define two Words types in `Int` and `Int64`, and have the rest of the types use typealiases according to the following table:

Type | Words
----- | -----
Int8 ... Int32 | Int.Words
UInt8 ... UInt32 | Int.Words
UInt32 | Int.Words
UInt64 | Int64.Words

So we would be left with two custom `Words` collection types (instead of 10), with consistent definitions on all platforms.

moiseev commented:
While reviewing your last commit, I noticed that `CollectionOfOne` conforms to `MutableCollection`, so effectively `var words` for all built-in integer types will be mutable.

moiseev commented:
Well, the `_lowWord` problem can be solved by a `BuiltinInteger` marker protocol, or something along these lines. This is not the first time when something like this sounds like a good idea.

lorentey commented:
Hm, mutability seems like an argument against `CollectionOfOne` here, as it adds another inconsistency with `Int64.Words` on 32-bit platforms. (Unless we make that mutable, too, of course.) Keeping a custom collection for even the single-word case is looking much more attractive to me now.

lorentey commented:
Even though the `words` property is read-only, mutability still seems like a concern for code like this:

```
var words = (42 as UInt64).words
words[0] = 23
```

This would compile on 64 bit platforms but not on 32-bit ones. (I'm not sure why anyone would actually write code like that, but it seems better to rule it out.)

lorentey committed:
[stdlib] Restore consistency between 32- and 64-bit platforms.

Don’t use CollectionOfOne for single-word cases; keep defining
custom collections for UInt.Words and UInt64.Words.

On 32-bit platforms, UInt64.Words and UInt.Words need to be
different types. For consistency, the two are now kept separate even
on 64-bit platforms.

Int64.Words is defined to be a typealias for UInt64.Words; all other
integer types use UInt.Words for their words type. This way, Words
is always defined the same way on all platforms, preventing a potential
subtle gotcha in user code.

UInt.Words could be a typealias for CollectionOfOne, but that would
make the asymmetry between [U]Int and [U]Int64 even more pronounced.
(E.g., CollectionOfOne is a mutable collection.)

lorentey commented:
Okay, I replaced the `DefaultWords` commit with one where we have just two custom collections, `UInt.Words` and `UInt64.Words`, with the rest of the integer types using a typealias to one of these two.

I decided against using `CollectionOfOne` for the `UInt.Words` case. While it would be nice to get rid of one of the two remaining collections, there are two major differences between `CollectionOfOne` and the custom collection `UInt64.Words`: `CollectionOfOne` is mutable and it has a custom iterator type. Neither feature seems important enough to emulate in `UInt64.Words`, but the asymmetry was nagging me. (Note that the custom iterator won't get much use in a single-word integer, as conversions between standard integer types will in most cases use a fast path involving `_lowWord`, rather than going through the `Words` collection. Furthermore, conversions to user-defined integer types can implement a `bitWidth` check followed by a direct conversion to `UInt` to achieve the same optimization.)

Latest stats: Compile time is 118.40 seconds (down from 121.06 on master), while libswiftCore.dylib is now 7,815,360 bytes (down 2.2% (by 169 KiB) from master's 7,988,400).

mastaeza reviewed (commented):
Ok

lorentey commented:
This is now mostly obsolete; `BinaryInteger.Words` is now officially constrained to `RandomAccessCollection`.
